### PR TITLE
Modernize Python versions, fix tests, and fix file leaks.

### LINKED ISSUE #1: Main leaks open files
The write to `~/.zope.teststats` (which is undocumented) isn't ever closed. This leads to dozens of warnings from, e.g., zopetoolkit.

https://github.com/zopefoundation/z3c.recipe.compattest/blob/34123bc9e2eef902e58e98dd63889757df6ce2cb/src/z3c/recipe/compattest/runner.py#L118-L127

### LINKED ISSUE #2: Python versions outdated

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #1 and fixes #2

The tests were very flaky because a lot of dependencies have added new extras since they were written, I think, and those extras differ from platform to platform (including across python versions). And the iteration order varied too, so it was always a different missing dependency. The fix was two parts:

- Don't install `zc.buildout` extras, which was the biggest source of the problem. I don't anticipate that being an issue.
- Filter the set of extras we install. This was because setuptools has a bunch of extras now. I could imagine there's a chance that could be an issue, but I don't *think* it is? If it is, we can add an option to not filter the extras.

### CODE REVIEW

File: src/z3c/recipe/compattest/runner.py [hunk 1]
```diff
@@ -48,17 +49,22 @@ def start(self):
...
-        self.exitcode = self.process.poll()
-        if self.exitcode is not None:
+        if self.process is not None:
+            self.exitcode = self.process.poll()
+        if self.exitcode is not None and self.process is not None:
             self.end = time.time()
             # We're done, get it all
             data = self.process.stdout.read()
+            # Close the pipes and cleanup.
+            self.process.communicate()
```
Comment (mgedmin) [inline src/z3c/recipe/compattest/runner.py:63]: Why not combine the reading/closing into one statement, namely

```python
data = self.process.communicate()[0]
```

?

**Review by mgedmin — APPROVED** [review]
LGTM

File: src/z3c/recipe/compattest/runner.py (hunk 1 above)
Comment (jamadden) [reply src/z3c/recipe/compattest/runner.py:63]: Good idea, thanks.

Comment (jamadden) [reply src/z3c/recipe/compattest/runner.py:63]: Apparently [that breaks PyPy3-7.1](https://travis-ci.org/github/zopefoundation/z3c.recipe.compattest/jobs/672912967). There seems to be a buffering issue combining direct reads and the `communicate()` method.

**Comment by jamadden** [comment]
When this is merged, I'd like to make a release of this. Could one of the PyPI maintainers please grant me (`jamadden`) and @mgedmin (`mgedmin`) PyPI rights? [The PyPI maintainers](https://pypi.org/project/z3c.recipe.compattest/) appear to include @janjaapdriessen @jimfulton @hannosch @janwijbrand @ctheune @thefunny42 .

**Review by mgedmin — APPROVED** [review]
LGTM

File: src/z3c/recipe/compattest/runner.py (hunk 1 above)
Comment (mgedmin) [reply src/z3c/recipe/compattest/runner.py:63]: I don't suppose you've [filed a PyPy bug](https://bitbucket.org/pypy/pypy/issues)?

Comment (jamadden) [reply src/z3c/recipe/compattest/runner.py:63]: It's already fixed; the version on Travis in outdated.

Comment (mgedmin) [reply src/z3c/recipe/compattest/runner.py:63]: Do you know which pypy3 version fixed that bug?

`dist: bionic` on Travis currently has pypy3 7.2.0.

Actually it doesn't matter; the workaround is simple.

Comment (jamadden) [reply src/z3c/recipe/compattest/runner.py:63]: It works for me with pypy3 7.3.0.

**Pull request merged**